doc: Add ADR 004-Component-architecture (#6)
nixops4/nixops4: doc/architecture/004-Component-architecture.md

# Component architecture

## Context

As previously discussed, NixOps 4's architecture has great similarity with Nix's architecture.

The goal of this document is to outline the components and layers that make up NixOps 4.

## Decision

At a high level, `nixops4` is an executable that realises its extensions through Nix, starts them, and interacts with them.

`nixops4` individually has a layered application architecture, with the following layers:

- The NixOps CLI, analogous to Nix's `src/nix`
  - Implements the command line interface
- The NixOps evaluator process
  - Implements an interface provided by the NixOps core library
  - Exists as a separate process, to improve the robustness of the system

>     - Data loss following resource creation can have a real world cost, especially if not reported to the user.

Reviewer: Good to make it explicit that this is because the evaluator is more likely to crash. It interfaces with a C library, so there's less control over it crashing or corrupting data.

- The NixOps core library, analogous to Nix's `libstore`, and using it, but not `libexpr`.

>   - Implements the interaction with resource implementations

Author: Define _resource_

Author: https://github.com/nixops4/nixops4/pull/6#discussion_r1699660087

>   - Implements the interaction with resource implementations

Author: ```suggestion
  - Implements the interaction with resource implementations
    - for context, a resource represents a "real world" object, like a DNS record or cloud VM instance, and tends to implement CRUD operations. The Terraform/OpenTofu definition of resource is a good intuition for what a NixOps resource will be
```

>   - Coordinates the interactions between resources and the evaluator

Author: ```suggestion
  - Coordinates the interactions between resource processes and the evaluator process
```

> 

Reviewer: I'm not left with a clear picture of what the library contains. Based on our conversation:
- bulk of the code that the CLI runs
- code for the evaluator to interact with the CLI

> 

Author: ```suggestion
  - Maintains a model of the configuration
  - Invokes operations on resources (process creation and communication)
  - The bulk of the logic exposed through the CLI, but not the CLI details such as flag names, etc

```
